### DISCUSSION TITLE
Extending Order with custom fields

### DISCUSSION THREAD

**Discussion by vosco88** [OFFLINE-GmbH/oc-mall-plugin#1058] [Q&A] [answered]
Hello, 

would there be a way to extend the order model during checkout with custom fields. 

For example:
- note from the user
- phone for delivery companies - as each order can have different phone so it would be cumbersome to go through addresses each time for users
- gift orders notes
etc.

Thank you

**Comment by tobias-kuendig** [comment]
Everything in October is highly extendable. You can extend Mall's Order model and form according to the docs:
https://docs.octobercms.com/3.x/extend/extending.html#extending-a-user-model

You can also write migrations that add custom columns to the Order table or store the additional data in a custom table with a foreign key to the order.

**Reply by vosco88** [reply]
Thank you I got this part. 

I am just not sure how would I do it within the boundaries of QuickCheckout component. 

- I've added phone number field to the quick checkout component

When I check the onSubmit of the component, where could I hook into the post data to get this phone number and possibly write it to cart for example so then on mall.order.beforeCreate I can move it to the Order model. 
Would I possibly need to write my own component extending QuickCheckout or something like that?

The idea was to do it outside - so when user fills in the phone number I would write it to cart and then use the event to update the cart with it. I am not sure that this is a good solution as it introduces more javascript. 

Thank you for the help.

**Reply by tobias-kuendig** [reply, marked as answer]
>  Would I possibly need to write my own component extending QuickCheckout or something like that?

Absolutely! Another option is to use the beforeCreate model event of the Order model and check the context. Is the current request a ajax request? Is it on your checkout URL? Is your custom data available via `post()`? If all is true, then set it on the model.

**Reply by vosco88** [reply]
Second option worked, the added field was in the post() so beforeCreate on model works. 

One follow up question where would I validate the added field within the flow? Extending the rules on Order model?

**Reply by tobias-kuendig** [reply]
> Extending the rules on Order model? 

You can do that or there is a beforeValidate event you can use to run additional validation logic.

**Comment by ogalbnafets** [comment]
Hello vesco88,

could you provide an example how you extended the Orders? I'm also trying to add a order comment but something I'm missing... I'm not experiencing in plugin development.
Thank you very much!

**Reply by adamo** [reply]
I'd love to know the same - my struggle is to validate the custom fields for a custom shipping method. If the method is selected then the user has to fill out shipping details for this particular order(pickup point id).
I've tried to use the `beforeValidate`, `beforeCreate`, adding a rule to the required fields on the order model but with no luck.

My problem is that I'd like to validate the custom order field, but even if the validation is working (for example forcing it by adding it to $required fields) the user is already created before the validation fires and the next error is `this email is taken`.

**Reply by ogalbnafets** [reply]
Hello Adam!
I posted my solution in the other
thread:
https://github.com/OFFLINE-GmbH/oc-mall-plugin/discussions/1103#discussioncomment-8806561
FYI: I added also to the Delivery Model a field 'is_takeaway' - and when in the quickqueckout a product is found which is deactivated for delivery and the chosen delivery method has not is_takeaway it will show an error - maybe you need this? Message me if you want!
Best regards

**Reply by adamo** [reply]
Thanks for the response @ogalbnafets the is_takeaway look as something that might be helpful so I'd appreciate it. 
My current problem is that I need to make the field added to order required, but when using validation first the user is validated, and then if the users passes validation my custom fields - by this time the user is created and when the customer fills out my field, the next validation error is: this email is taken.
